### PR TITLE
Graceful handling and notification of errors

### PR DESCRIPTION
**Pull request by AshishA26**
worker_manager.py has been updated to recover from a worker crashing. Code checks if all workers are alive in each manager. If a worker is dead, log the error and restart the worker (terminate and join the worker, drain the input queues, and create and start a new worker).

### CODE REVIEW

File: utilities/workers/worker_manager.py [hunk 1]
```diff
@@ -106,6 +111,18 @@ def get_worker_target(self) -> "(...) -> object":  # type: ignore
...
+        """
+        Returns the input queues.
+        """
+        return self.__input_queues
+
+    def get_worker_name(self) -> str:
+        """
+        Returns the name of the worker type.
+        """
+        return self.__worker_name
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:124]: Is it possible to get the name of the target function instead?

File: utilities/workers/worker_manager.py [hunk 2]
```diff
@@ -195,3 +218,35 @@ def join_workers(self) -> None:
...
+                # Terminate and join worker
+                worker.terminate()
+                worker.join()
+                del worker
+                worker = None
+
+                # Drain the preceding queues
+                input_queues = self.__worker_properties.get_input_queues()
+                for queue in input_queues:
+                    queue.drain_queue()
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:242]: It is inefficient to drain the queues every time the loop encounters a dead worker. Pull this out of the loop so that it only needs to be done once.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: utilities/workers/worker_manager.py (hunk 1 above)
Comment (AshishA26) [reply utilities/workers/worker_manager.py:124]: We would have to create getter in each worker I think. Would this be a good idea, I'm not sure?

Comment (Xierumeng) [reply utilities/workers/worker_manager.py:124]: I tried this and it worked:
```py
def test():
    pass


if __name__ == "__main__":
    a = test
    print(a.__name__)
```
So I think you could have:
```py
def get_target_name(self) -> str:
    return self.__target.__name__
```

Comment (Xierumeng) [reply utilities/workers/worker_manager.py:124]: Actually, you can directly use the worker's process name instead:
```py
for worker in self.__workers:
    print(worker.name)
```
Source: https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process.name

File: utilities/workers/worker_manager.py [hunk 3]
```diff
@@ -195,3 +218,35 @@ def join_workers(self) -> None:
...
+                # Log the error
+                frame = inspect.currentframe()
+                worker_name_string = self.__worker_properties.get_worker_name() + " worker"
+                self.__local_logger.error("Worker died, restarting " + worker_name_string, frame)
+
+                # Terminate and join worker
+                worker.terminate()
+                worker.join()
+                del worker
+                worker = None
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:237]: This seems sketchy, considering that failing to create a worker below leaves this in the list. I think it's better to copy the alive workers into a new list, and then create and append whatever number of workers is required to bring it back to the original size.

File: utilities/workers/worker_manager.py [hunk 4]
```diff
@@ -195,3 +218,35 @@ def join_workers(self) -> None:
...
+
+                # Create a new worker
+                result, worker = WorkerManager.__create_single_worker(
+                    self.__worker_properties.get_worker_target(),
+                    self.__worker_properties.get_worker_arguments(),
+                    self.__local_logger,
+                )
+                if not result:
+                    frame = inspect.currentframe()
+                    self.__local_logger.error("Failed to restart " + worker_name_string, frame)
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:247]: This should return something so that the caller knows that there is a catastrophic failure.

File: utilities/workers/worker_manager.py [hunk 5]
```diff
@@ -195,3 +218,35 @@ def join_workers(self) -> None:
...
+
+        # Drain the preceding queues
+        input_queues = self.__worker_properties.get_input_queues()
+        for queue in input_queues:
+            queue.drain_queue()
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:252]: Hmm, why is this necessary?

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: utilities/workers/worker_manager.py (hunk 5 above)
Comment (AshishA26) [reply utilities/workers/worker_manager.py:252]: Draining the preceding queue ensures that the preceding queue data wasn't what caused the worker to fail. As for draining the succeeding queues, I don't think it is necessary because a worker that died wouldn't have put bad data into the queue.

File: utilities/workers/worker_manager.py (hunk 1 above)
Comment (AshishA26) [reply utilities/workers/worker_manager.py:124]: using worker.name would print out "Worker died, restarting Process-5". I'm not sure how useful this would be, should we be concatenating worker.name with the target name?

Comment (Xierumeng) [reply utilities/workers/worker_manager.py:124]: That would work.

File: log.txt
Comment (Xierumeng) [inline log.txt:1]: Remove.

File: utilities/workers/worker_manager.py [hunk 6]
```diff
@@ -219,34 +219,43 @@ def join_workers(self) -> None:
...
+        Returns whether the dead workers were able to be restarted.
         """
+        new_workers = []
         for worker in self.__workers:
-            if not worker.is_alive():
-                # Log the error
+            if worker.is_alive():
+                new_workers.append(worker)
+
+        del self.__workers
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:233]: I don't think this is necessary.

File: utilities/workers/worker_manager.py [hunk 7]
```diff
@@ -195,3 +218,44 @@ def join_workers(self) -> None:
...
+        """
+        new_workers = []
+        for worker in self.__workers:
+            if worker.is_alive():
+                new_workers.append(worker)
+
+        del self.__workers
+        self.__workers = new_workers
+
+        while len(self.__workers) < self.__worker_properties.get_worker_count():
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:236]: Use a for loop instead.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: utilities/workers/worker_manager.py (hunk 5 above)
Comment (Xierumeng) [reply utilities/workers/worker_manager.py:259]: Add a comment.

File: utilities/workers/worker_manager.py [hunk 8]
```diff
@@ -195,3 +213,49 @@ def join_workers(self) -> None:
...
+        """
+        Check and restart dead workers.
+
+        Returns whether the dead workers were able to be restarted.
+        """
+        new_workers = []
+        for worker in self.__workers:
+            if worker.is_alive():
+                new_workers.append(worker)
+            else:
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:227]: Nit: Instead of else block:
```py
if worker.is_alive():
    new_workers.append(worker)
    continue

...
```

File: utilities/workers/worker_manager.py [hunk 9]
```diff
@@ -195,3 +213,49 @@ def join_workers(self) -> None:
...
+                # Log the error
+                frame = inspect.currentframe()
+                target_and_worker_name = (
+                    self.__worker_properties.get_target_name() + " " + worker.name
+                )
+                self.__local_logger.error(
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:233]: I think warning or info level is more appropriate, because error level implies almost unrecoverable failure.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: utilities/workers/worker_manager.py [hunk 10]
```diff
@@ -195,3 +213,48 @@ def join_workers(self) -> None:
...
+
+        Returns whether the dead workers were able to be restarted.
+        """
+        new_workers = []
+        for worker in self.__workers:
+            if worker.is_alive():
+                new_workers.append(worker)
+                continue
+
+            # Log the error
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:229]: Update this comment. For example: `# Log dead worker`

File: utilities/workers/worker_manager.py [hunk 11]
```diff
@@ -195,3 +213,48 @@ def join_workers(self) -> None:
...
+            frame = inspect.currentframe()
+            target_and_worker_name = self.__worker_properties.get_target_name() + " " + worker.name
+            self.__local_logger.warning(
+                "Worker died, restarting " + target_and_worker_name,
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:233]: Use f strings instead of string concatenation.

File: utilities/workers/worker_manager.py [hunk 12]
```diff
@@ -195,3 +213,48 @@ def join_workers(self) -> None:
...
+
+            # Create a new worker
+            result, new_worker = WorkerManager.__create_single_worker(
+                self.__worker_properties.get_worker_target(),
+                self.__worker_properties.get_worker_arguments(),
+                self.__local_logger,
+            )
+            if not result:
+                frame = inspect.currentframe()
+                self.__local_logger.error("Failed to restart " + target_and_worker_name, frame)
```
Comment (Xierumeng) [inline utilities/workers/worker_manager.py:245]: f strings.

**Review by Xierumeng — COMMENTED** [review]
Reviewd.

**Review by Xierumeng — APPROVED** [review]
Approved.
